### PR TITLE
Implementation of syscalls in bpf_syscall.cpp

### PR DESCRIPTION
**Pull request by delaramamiri**
## Description
Add support for the following APIs:
```
BPF_PROG_ATTACH
BPF_PROG_DETACH
BPF_MAP_LOOKUP_AND_DELETE_ELEM
```
List of all cilium syscalls [is included here:](https://github.com/cilium/ebpf/blob/71b98d0977c1d669c709ff6ce897e5f356d06aec/internal/sys/types.go#L67)

## Testing
Added unit tests for the aforementioned APIs in  unit tests. 

## Documentation

No 

Resolves issue #1824

### CODE REVIEW

**Review by Alan-Jowett — DISMISSED** [review]

File: include/linux/bpf.h [hunk 1]
```diff
@@ -87,6 +90,15 @@ typedef struct
     uint32_t link_fd; ///< File descriptor of link to detach.
 } bpf_link_detach_attr_t;
 
+/// Attributes used by BPF_PROG_ATTACH/DETACH .
+typedef struct
+{
+    uint32_t prog_fd;       ///< File descriptor of program to attach to.
+    uint32_t target_fd; ///< eBPF target to attach/detach to/from.
+    enum bpf_attach_type prog_type; ///< Type of program to attach/detach to/from.
+    uint32_t flags; ///< Flags affecting the attach operation.
```
Comment (dthaler) [inline include/linux/bpf.h:99]: nit: Either make all these comments line up, or make all comments only have 1 space (line 96 doesn't).  I think the former is what the code formatter usually does.

File: tests/unit/libbpf_test.cpp [hunk 2]
```diff
@@ -2498,6 +2498,21 @@ TEST_CASE("BPF_PROG_BIND_MAP etc.", "[libbpf]")
     REQUIRE(map_fd2 > 0);
     Platform::_close(map_fd2);
 
+    //Verify we attach the program.
+    memset(&attr, 0, sizeof(attr));
+    attr.bpf_prog.prog_fd= program_fd;
+    attr.bpf_prog.target_fd = program_fd;
+    attr.bpf_prog.flags = 0;
+    attr.bpf_prog.prog_type = BPF_CGROUP_INET4_CONNECT;
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2506]: I see you changed this from XDP to this.  Does it matter?  If it's not supposed to work with XDP, is there a negative test somewhere?

File: tests/unit/libbpf_test.cpp [hunk 3]
```diff
@@ -2510,9 +2510,58 @@ TEST_CASE("BPF_PROG_BIND_MAP etc.", "[libbpf]")
...
+
+    struct bpf_program* program = bpf_object__find_program_by_name(object, "authorize_connect4");
+    REQUIRE(program != nullptr);
+
+    REQUIRE(bpf_object__load(object) == 0);
+
+    int program_fd = bpf_program__fd(program);
+    REQUIRE(program_fd > 0);
+
+    //// Verify we can't attach the program to an attach type that doesn't work with this API.
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2532]:
```suggestion
    // Verify we can't attach the program using an attach type that doesn't work with this API.
```

File: tests/unit/libbpf_test.cpp [hunk 4]
```diff
@@ -2510,9 +2510,58 @@ TEST_CASE("BPF_PROG_BIND_MAP etc.", "[libbpf]")
...
+    memset(&attr, 0, sizeof(attr));
+    attr.bpf_prog.prog_fd = program_fd;
+    attr.bpf_prog.target_fd = program_fd;
+    attr.bpf_prog.flags = 0;
+    attr.bpf_prog.prog_type = BPF_XDP;
+    REQUIRE(bpf(BPF_PROG_ATTACH, &attr, sizeof(attr)) == -ENOTSUP);
+
+    // Verify we attach the program.
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2540]:
```suggestion
    // Verify we can attach the program.
```

File: tests/unit/libbpf_test.cpp [hunk 5]
```diff
@@ -2510,9 +2510,58 @@ TEST_CASE("BPF_PROG_BIND_MAP etc.", "[libbpf]")
...
+    memset(&attr, 0, sizeof(attr));
+    attr.bpf_prog.prog_fd = program_fd;
+    attr.bpf_prog.target_fd = program_fd;
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2543]:
```suggestion
    // TODO (issue #1028): Currently the target_fd is treated as a compartment id.
    attr.bpf_prog.target_fd = program_fd;
```

File: tests/unit/libbpf_test.cpp [hunk 6]
```diff
@@ -2510,9 +2510,58 @@ TEST_CASE("BPF_PROG_BIND_MAP etc.", "[libbpf]")
...
+    attr.bpf_prog.flags = 0;
+    attr.bpf_prog.prog_type = BPF_CGROUP_INET4_CONNECT;
+    REQUIRE(bpf(BPF_PROG_ATTACH, &attr, sizeof(attr)) == 0);
+
+
+    // Verify we detach the program.
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2549]:
```suggestion
    // Verify we can detach the program.
```

File: tests/unit/libbpf_test.cpp [hunk 7]
```diff
@@ -2510,9 +2510,58 @@ TEST_CASE("BPF_PROG_BIND_MAP etc.", "[libbpf]")
...
+    memset(&attr, 0, sizeof(attr));
+    attr.bpf_prog.target_fd = program_fd;
+    attr.bpf_prog.prog_type = BPF_CGROUP_INET4_CONNECT;
+    REQUIRE(bpf(BPF_PROG_DETACH, &attr, sizeof(attr)) == 0);
+
+   // Verify we can't detach= the program from a type that doesn't work with this API.
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2555]:
```suggestion
   // Verify we can't detach the program from a type that doesn't work with this API.
```

File: tests/unit/libbpf_test.cpp [hunk 8]
```diff
@@ -2562,13 +2611,24 @@ TEST_CASE("BPF_MAP_GET_NEXT_KEY etc.", "[libbpf]")
...
     REQUIRE(errno == ENOENT);
-
+  
     // Delete the entry.
     memset(&attr, 0, sizeof(attr));
     attr.map_fd = map_fd;
     attr.key = (uintptr_t)&key;
     REQUIRE(bpf(BPF_MAP_DELETE_ELEM, &attr, sizeof(attr)) == 0);
 
+    // Lookup and delete the entry.
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2621]:
```suggestion
    // Look up and delete the entry.
```

File: tests/unit/libbpf_test.cpp [hunk 9]
```diff
@@ -2562,13 +2611,24 @@ TEST_CASE("BPF_MAP_GET_NEXT_KEY etc.", "[libbpf]")
...
+    memset(&attr, 0, sizeof(attr));
+    value = 0;
+    key = 42;
+    attr.map_fd = map_fd;
+    attr.key = (uintptr_t)&key;
+    attr.value = (uintptr_t)&value;
+    //Add the element back to the entry after the previous test entry deletion.
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2628]:
```suggestion

    // Add the element back to the entry after the previous test entry deletion.
```

**Review by dthaler — COMMENTED** [review]
Looks fine other than cleaning up some // comments.

File: include/linux/bpf.h [hunk 10]
```diff
@@ -87,6 +90,15 @@ typedef struct
...
 } bpf_link_detach_attr_t;
 
+/// Attributes used by BPF_PROG_ATTACH/DETACH .
+typedef struct
+{
+    uint32_t prog_fd;               ///< File descriptor of program to attach to.
+    uint32_t target_fd;             ///< eBPF target to attach/detach to/from.
+    enum bpf_attach_type prog_type; ///< Type of program to attach/detach to/from.
+    uint32_t flags;                 ///< Flags affecting the attach operation.
+} bpf_prog_attr_t;
```
Comment (dthaler) [inline include/linux/bpf.h:100]: This doesn't seem to match libbpf.  The fields on Linux are called:
```
        uint32_t target_fd;
        uint32_t attach_bpf_fd;
        uint32_t attach_type;
        uint32_t attach_flags;
        uint32_t replace_bpf_fd;
```

File: include/linux/bpf.h [hunk 11]
```diff
@@ -173,6 +185,9 @@ union bpf_attr
     // BPF_LINK_DETACH
     bpf_link_detach_attr_t link_detach; ///< Attributes used by BPF_LINK_DETACH.
 
+    // BPF_PROG_ATTACH
+    bpf_prog_attr_t bpf_prog; ///< Attributes used by BPF_PROG_ATTACH and BPF_PROG_DETACH.
```
Comment (dthaler) [inline include/linux/bpf.h:189]: For cross-plat compat, this should be an unnamed struct, not a named struct.  See how BPF_PROG_LOAD is done at line 143.

File: tests/unit/libbpf_test.cpp [hunk 12]
```diff
@@ -2562,13 +2612,25 @@ TEST_CASE("BPF_MAP_GET_NEXT_KEY etc.", "[libbpf]")
...
+    value = 0;
+    key = 42;
+    attr.map_fd = map_fd;
+    attr.key = (uintptr_t)&key;
+    attr.value = (uintptr_t)&value;
+
+    // Add the element back to the entry after the previous test entry deletion.
+    bpf(BPF_MAP_UPDATE_ELEM, &attr, sizeof(attr));
+    REQUIRE(bpf(BPF_MAP_LOOKUP_AND_DELETE_ELEM, &attr, sizeof(attr)) == 0);
+
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2633]: add another check in here when the element doesn't exist.  I.e., call bpf(BPF_MAP_LOOKUP_AND_DELETE_ELEM, &attr, sizeof(attr)) again and test the return code.

File: tests/unit/libbpf_test.cpp [hunk 13]
```diff
@@ -2510,9 +2510,59 @@ TEST_CASE("BPF_PROG_BIND_MAP etc.", "[libbpf]")
...
+    // Verify we can attach the program.
+    memset(&attr, 0, sizeof(attr));
+    attr.attach_bpf_fd = program_fd;
+    // TODO (issue #1028): Currently the target_fd is treated as a compartment id.
+    attr.target_fd = program_fd;
+    attr.attach_flags = 0;
+    attr.attach_type = BPF_CGROUP_INET4_CONNECT;
+    REQUIRE(bpf(BPF_PROG_ATTACH, &attr, sizeof(attr)) == 0);
+
```
Comment (dthaler) [inline tests/unit/libbpf_test.cpp:2548]:
```suggestion
```
nit: remove extra blank line

**Review by dthaler — APPROVED** [review]
Approving but see suggestions.

**Review by dthaler — APPROVED** [review]

File: include/linux/bpf.h [hunk 14]
```diff
@@ -140,7 +143,17 @@ union bpf_attr
         uint32_t log_size;     ///< Size in bytes of the log buffer.
         uint32_t kern_version; ///< Kernel version (currently ignored on Windows).
     };                         ///< Attributes used by BPF_PROG_LOAD.
-
+    
+    //BPF_PROG_ATTACH
```
Comment (dv-msft) [inline include/linux/bpf.h:147]: Dead code?

**Review by dv-msft — APPROVED** [review]

**Pull request merged**